### PR TITLE
feat(content-server settings): add button for payments/subscription management to /settings

### LINKED ISSUE #924: Create /settings button to point to subscription management
This is a ticket to split off a portion of #716.

We should have a button on the `/settings` page that will send someone to the payments page. The link has to be dynamically generated to include an authentication token.

#926 is for creating an endpoint whose only purpose is redirecting.

We should also have a feature-flag based email whitelist, so that we can expose it only to `@mozilla.com` and `@softvision` users, to do testing in prod.

### PR DESCRIPTION
**Pull request by ianb**
This pulls out the button from PR #907, and the auth token generation from [Les's commit](https://github.com/lmorchard/fxa/commit/47e632b5020381082cdc3331f3398f4513e43ff8).

~There's some FIXMEs about enabling the feature, with a feature flag and possibly also a config setting; I didn't get a chance to do those, but I will be out Friday and wanted to get something up for feedback.~

@lmorchard took a shot at config & feature flags stuff here

Fixes #924

### CODE REVIEW

**Comment by lmorchard** [comment]
I went ahead and made some more tweaks on this along with a rebase against master. Took a shot at moving a bunch of things into config and added a feature flag for the management button.

File: packages/fxa-content-server/app/scripts/views/settings/subscription.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+  submit () {
+    const {
+      managementClientId,
+      managementScopes,
+      managementTokenTTL,
+      managementUrl,
+    } = this._config;
+    const account = this.user.getSignedInAccount();
+    account
```
Comment (lmorchard) [inline packages/fxa-content-server/app/scripts/views/settings/subscription.js]: Oh, also, wanted to note: This is not the final implementation for generating this token. Leaving #881 open to work out how to better generate and pass a token to the subscription management pages, probably involving a code exchange that enforces an IP address check to help ensure it's the same user navigating from FxA.

**Comment by lmorchard** [comment]
Okay, one more tweak to shift the button visibility logic into the parent view and add some tests.

File: packages/fxa-content-server/app/scripts/views/settings/subscription.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+'use strict';
+
+import Cocktail from 'cocktail';
+import FormView from '../form';
+import SettingsPanelMixin from '../mixins/settings-panel-mixin';
+import Template from 'templates/settings/subscription.mustache';
+
+const View = FormView.extend({
+  template: Template,
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/settings/subscription.js]: I know parts of this module are going to be changed, but can you add a unit test suite to ensure that it renders without blowing up?

File: packages/fxa-content-server/app/scripts/lib/app-start.js [hunk 3]
```diff
@@ -428,6 +428,18 @@ Start.prototype = {
   },
 
   createView (Constructor, options = {}) {
+    const {
+      subscriptions: {
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/lib/app-start.js]: Please remove the development config from code and rely upon the [passed in config from get-index.js](https://github.com/mozilla/fxa/pull/940/files#diff-da283abf354df26a8bebd2d236b18f39R55).

File: packages/fxa-content-server/app/scripts/lib/app-start.js [hunk 4]
```diff
@@ -428,6 +428,18 @@ Start.prototype = {
...
+        allowedLanguages: subscriptionsManagementLanguages = ['en-US'],
+        enabled: subscriptionsEnabled = false,
+      } = {},
+      featureFlags: {
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/lib/app-start.js]: I'd not worry about the featureFlags for this PR, do that in a follow on.

File: packages/fxa-content-server/app/scripts/views/settings/subscription.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  className: 'subscription',
+  viewName: 'settings.subscription',
+
+  events: {
+    'click button': 'submit',
+  },
+
+  initialize (options = {}) {
+    this._config = Object.assign({}, {
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/settings/subscription.js]: See the [comment above about removing the hard coded config](https://github.com/mozilla/fxa/pull/940/files#diff-0370004e39359bfe3201953deacddcacR432) here and instead relying upon the [config passed in from configuration.js](https://github.com/mozilla/fxa/pull/940/files#diff-e6c30fd24af1b7b4b7d351ad516d6092R597).

File: packages/fxa-content-server/app/scripts/views/settings/subscription.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      managementClientId: '98e6508e88680e1a',
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/settings/subscription.js]: This is the content server's client_id, let's use a value of `59cceb6f8c32317c` for the subscriptions page.

File: packages/fxa-content-server/server/lib/configuration.js [hunk 7]
```diff
@@ -594,6 +594,44 @@ const conf = module.exports = convict({
...
+      format: Array,
+    },
+    enabled: {
+      default: false,
+      doc: 'Indicates whether subscriptions APIs are enabled',
+      env: 'SUBSCRIPTIONS_ENABLED',
+      format: Boolean,
+    },
+    managementClientId: {
+      default: 'YOU MUST CHANGE ME',
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/server/lib/configuration.js]: Let's use a default of `59cceb6f8c32317c`. I'll open a PR with an entry on the oauth server.

File: packages/fxa-auth-server/config/dev.json [hunk 8]
```diff
@@ -75,6 +76,7 @@
       "exampleProd3": [ "exampleCap1", "exampleCap2", "exampleCap3", "exampleCap4", "exampleCap5" ]
     },
     "clientCapabilities": {
+      "98e6508e88680e1a": [ "isRegistered", "isSubscribed" ],
```
Comment (shane-tomlinson) [inline packages/fxa-auth-server/config/dev.json]: Should this be the FxA settings page client_id or the payments page? If the payment's page, please use a client_id of `59cceb6f8c32317c`

Comment (shane-tomlinson) [reply packages/fxa-auth-server/config/dev.json]: Is this logic actually used anywhere or should this be part of a separate PR?

File: packages/fxa-content-server/app/scripts/views/settings.js [hunk 9]
```diff
@@ -220,10 +224,31 @@ const View = BaseView.extend({
...
+   * Should the subscriptions management panel be displayed?
+   *
+   * @returns {Boolean}
+   * @private
+   */
+  _isSubscriptionsManagementVisible () {
+    if (! this._subscriptionsManagementEnabled) {
+      return false;
+    }
+    const acceptedLanguages = (this.navigator.languages || [])
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/settings.js]: The locale the rest of the FxA UI is being displayed in is defined in [config.lang](https://github.com/mozilla/fxa/blob/68978f34c9ed96536a6adc490f6d2f6a178d34b6/packages/fxa-content-server/app/scripts/lib/config-loader.js#L33), where [config is options.config in the initialize function](https://github.com/mozilla/fxa/pull/940/files#diff-33ca28c5444ea7de91653e837ef072caR70).

Is it true the only gate is locale, or is it the user's location, i.e., the country in which they are located? If it's the country, we don't have that info on the front end currently, but should be able to add it with minimal fuss, perhaps passed in with [serializedConfig in get-index](https://github.com/mozilla/fxa/pull/940/files#diff-da283abf354df26a8bebd2d236b18f39R55).

File: packages/fxa-content-server/server/lib/routes/get-index.js [hunk 10]
```diff
@@ -51,6 +52,7 @@ module.exports = function (config) {
     scopedKeysEnabled: SCOPED_KEYS_ENABLED,
     scopedKeysValidation: SCOPED_KEYS_VALIDATION,
     staticResourceUrl: STATIC_RESOURCE_URL,
+    subscriptions: SUBSCRIPTIONS,
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/server/lib/routes/get-index.js]: Please [update the unit test](https://github.com/mozilla/fxa/blob/68978f34c9ed96536a6adc490f6d2f6a178d34b6/packages/fxa-content-server/tests/server/routes/get-index.js#L79) to ensure this field and the expected values it contains are sent.

File: packages/fxa-shared/feature-flags/schema.json [hunk 11]
```diff
@@ -22,6 +22,14 @@
       "minimum": 0,
       "maximum": 1
     },
+    "subscriptions": {
```
Comment (shane-tomlinson) [inline packages/fxa-shared/feature-flags/schema.json]: Like I said above, I'd leave the feature flag to another PR for simplicity.

File: packages/fxa-content-server/app/scripts/lib/app-start.js (hunk 4 above)
Comment (lmorchard) [reply packages/fxa-content-server/app/scripts/lib/app-start.js]: I added this because we talked in last retro about starting to use feature flags for unfinished stuff like this.

What would the follow-on PR look like? Wondering if I'm missing things for the feature flag that would complicate it.

File: packages/fxa-auth-server/config/dev.json (hunk 8 above)
Comment (lmorchard) [reply packages/fxa-auth-server/config/dev.json]: I think is spillover from the demo I threw together this past Monday, I'll just remove.

File: packages/fxa-content-server/app/scripts/lib/app-start.js (hunk 4 above)
Comment (shane-tomlinson) [reply packages/fxa-content-server/app/scripts/lib/app-start.js]: > I added this because we talked in last retro about starting to use feature flags for unfinished stuff like this.

Yeah, understood. One of the main intentions is so we can modify the config mid-train w/o requiring a re-spin of the stack. Normal config requires a full re-spin w/ the new config, feature flagging no. I'm assuming this won't ship the feature mid-train, at least not in the train this lands on, so I'm suggesting pulling it out for now, because...

> What would the follow-on PR look like?

Because I don't actually know what the feature flagging code should look like and don't want to hold up this PR while trying to figure it out. :/

File: packages/fxa-content-server/app/scripts/views/settings.js (hunk 9 above)
Comment (lmorchard) [reply packages/fxa-content-server/app/scripts/views/settings.js]: I think so far, it's only locale because the payment pages won't be localized at launch. But, I think we're going to have a follow-up to this that will also hide the button if the user doesn't have any active subscriptions - which, itself, will be gated by geographic market.

Comment (lmorchard) [reply packages/fxa-content-server/app/scripts/views/settings.js]: Hmm, looking at config.lang, that's just language and not locale. That's probably good enough here for now, because along with geographic market I think we're not applying l10n to the payment pages for MVP

File: packages/fxa-content-server/server/lib/configuration.js (hunk 7 above)
Comment (lmorchard) [reply packages/fxa-content-server/server/lib/configuration.js]: I used `98e6508e88680e1a` here just to get it working, pending a resolution for #881 to introduce the real OAuth client

File: packages/fxa-content-server/app/scripts/lib/app-start.js (hunk 4 above)
Comment (lmorchard) [reply packages/fxa-content-server/app/scripts/lib/app-start.js]: Okie doke! I also just remembered that `SUBSCRIPTIONS_ENABLED` in the overall config defaults to false and also acts to hide this button. So, it's practically a feature flag, too - albeit a deploy-controlled one. But, I think that's fine for this.

**Comment by lmorchard** [comment]
Okay, force-pushed with tweaks in response to review by @shane-tomlinson, squashed & rebased against master.

**Comment by lmorchard** [comment]
Oh hey, I somehow missed that test failure locally. Pushed a tweak.

File: packages/fxa-content-server/server/lib/configuration.js [hunk 12]
```diff
@@ -594,6 +594,44 @@ const conf = module.exports = convict({
...
+    },
+    managementClientId: {
+      default: '59cceb6f8c32317c',
+      doc: 'OAuth client ID for subscriptions management pages',
+      env: 'SUBSCRIPTIONS_MANAGEMENT_CLIENT_ID',
+      format: String,
+    },
+    managementScopes: {
+      default: 'profile https://identity.mozilla.com/account/subscriptions',
+      doc: 'The publicly visible URL of the subscription page/service',
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/server/lib/configuration.js]: This doc seems incorrect.

File: packages/fxa-content-server/app/scripts/views/settings.js [hunk 13]
```diff
@@ -220,10 +225,29 @@ const View = BaseView.extend({
...
+   * Should the subscriptions management panel be displayed?
+   *
+   * @returns {Boolean}
+   * @private
+   */
+  _isSubscriptionsManagementVisible () {
+    if (! this._subscriptionsManagementEnabled) {
+      return false;
+    }
+    if (!this._subscriptionsManagementLanguages.includes(this._language)) {
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/settings.js]: nit: can you add a space after the `!`?

**Comment by lmorchard** [comment]
(and I just noticed I broke linting again ugh)

File: packages/fxa-content-server/server/lib/configuration.js [hunk 14]
```diff
@@ -594,6 +594,44 @@ const conf = module.exports = convict({
...
+      env: 'SUBSCRIPTIONS_MANAGEMENT_SCOPES',
+      format: String,
+    },
+    managementTokenTTL: {
+      default: 900,
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/server/lib/configuration.js]: 900 seconds is 15 minutes, is that the expected value? I thought I heard 5 minutes, but this can be reduced.

File: packages/fxa-content-server/app/scripts/views/settings.js (hunk 13 above)
Comment (shane-tomlinson) [reply packages/fxa-content-server/app/scripts/views/settings.js]: This was resolved.

Comment (lmorchard) [reply packages/fxa-content-server/app/scripts/views/settings.js]: Heh, yeah, I noticed a lint failure here pretty much 2 seconds before this comment 😅

File: packages/fxa-content-server/server/lib/configuration.js (hunk 14 above)
Comment (lmorchard) [reply packages/fxa-content-server/server/lib/configuration.js]: Hmm, I thought I heard 15 minutes... but IMO we can nail this down as a part of #881

Comment (shane-tomlinson) [reply packages/fxa-content-server/server/lib/configuration.js]: Yeah, merge this thing.

File: packages/fxa-auth-server/fxa-oauth-server/config/dev.json [hunk 15]
```diff
@@ -163,6 +163,18 @@
       "trusted": true,
       "allowedScopes": "https://identity.mozilla.com/apps/oldsync",
       "publicClient": true
+    },
```
Comment (shane-tomlinson) [inline packages/fxa-auth-server/fxa-oauth-server/config/dev.json]: If you rebase against master, I think this will already be there.
